cmd/action/migrate.go

```diff
@@ -0,0 +1,179 @@
+			}
+		},
+	}
+	// MigrateDiffCmd represents the migrate diff command.
+	MigrateDiffCmd = &cobra.Command{
+		Use:   "diff",
+		Short: "Compute the diff between the migration directory and a connected database and create a new migration file.",
+		Long: "`atlas migrate diff`" + ` uses the dev-database to re-run all migration files in the migration
+directory and compares it to a given desired state and create a new migration file containing SQL statements to migrate 
+the migration directory state to the desired schema. The desired state can be another connected database or an HCL file.`,
+		Example: `  atlas migrate diff --dev-url mysql://user:pass@localhost:3306/dev --to mysql://user:pass@localhost:3306/dbname
+  atlas migrate diff --dev-url mysql://user:pass@localhost:3306/dev --to file://atlas.hcl`,
+		Args: cobra.MaximumNArgs(1),
+		Run:  CmdMigrateDiffRun,
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(MigrateCmd)
+	MigrateCmd.AddCommand(MigrateDiffCmd)
+	MigrateCmd.PersistentFlags().StringVarP(&MigrateFlags.DirURL, migrateFlagDir, "", "file://migrations", "select migration directory using DSN format")
+	MigrateCmd.PersistentFlags().StringSliceVarP(&MigrateFlags.Schemas, migrateFlagSchema, "", nil, "set schema names")
+	MigrateCmd.PersistentFlags().BoolVarP(&MigrateFlags.Force, migrateFlagForce, "", false, "force a command to run on a broken migration directory state")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.DevURL, migrateDiffFlagDevURL, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.ToURL, migrateDiffFlagTo, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().SortFlags = false
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagDevURL))
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagTo))
+}
+
+func CmdMigrateDiffRun(cmd *cobra.Command, args []string) {
+	// Open a dev driver.
+	dev, err := DefaultMux.OpenAtlas(MigrateFlags.DevURL)
+	cobra.CheckErr(err)
+	// Open the migration directory. For now only local directories are supported.
+	dir, err := dir()
+	cobra.CheckErr(err)
+	// Get a state reader for the desired state.
+	desired, err := to(cmd.Context(), dev)
+	cobra.CheckErr(err)
+	// Only create one file per plan.
+	tfmt, err := migrate.NewTemplateFormatter(nameTmpl, contentTmpl)
+	cobra.CheckErr(err)
+	// Plan the changes and create a new migration file.
+	pl := migrate.NewPlanner(dev, dir, migrate.WithFormatter(tfmt))
+	var n string
+	if len(args) > 0 {
+		n = args[0]
+	}
+	plan, err := pl.Plan(cmd.Context(), n, desired)
+	cobra.CheckErr(err)
+	return
+	// Write the plan to a new file.
+	cobra.CheckErr(pl.WritePlan(plan))
+}
+
+// dir returns a migrate.Dir to use as migration directory. For now only local directories are supported.
+func dir() (migrate.Dir, error) {
```

Comment:
If the `dir` is generic, don't you think it's preferred to name it as `from`?

Reply:
We already have a `--from` flag in the `atlas schema diff` command and that is something else than here. Therefore I chose `--dir` to distinguish it from each other.